File: LeetCode/PHP/Arrays & Hashing/Easy/valid-anagram.php
<?php

/* Question Link: https://leetcode.com/problems/valid-anagram/ */

class Solution {

    /**
     * @param String $s
     * @param String $t
     * @return Boolean
     */
    function isAnagram($s, $t) {
        if(strlen($s) !== strlen($t)){
            return false;
        }

        $assocArr = [];

        for( $i = 0; $i < strlen($s) ; $i++){
            if(array_key_exists($s[$i], $assocArr)){
                $assocArr[$s[$i]] += 1;
            } else{
                $assocArr[$s[$i]] = 1;
            }
        }

        for( $i = 0; $i < strlen($t) ; $i++){
            if(!array_key_exists($t[$i], $assocArr) || $assocArr[$t[$i]] === 0){
                return false;
            }

            $assocArr[$t[$i]] -= 1;
        }

        return true;
    }
}

File: LeetCode/Python/Arrays & Hashing/Easy/valid-anagram.py
# Question Link: https://leetcode.com/problems/valid-anagram/

class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        # we can use a dictionary to store each char of s and how many times it has appeared
        # we can then run a for loop over t and check if that char exists in s's dictionary
        # if the char exists, we can reduce its value by 1 and if the key exists but its value is 0,
        # it means that char is greater in t than in s which means these two strings are not anagrams

        if len(s) != len(t):
            return False

        sDictionary = {}

        for char in s:
            if char in sDictionary:
               sDictionary[char] += 1
            else:
               sDictionary[char] = 1

        for char in t:
            if  char not in sDictionary or sDictionary.get(char) == 0:
                return False
            else:
                sDictionary[char] -= 1

        return True

        